engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/CatchPanicNode.java

```diff
@@ -54,24 +61,28 @@ Object doExecute(
     } catch (PanicException e) {
       panicBranchProfile.enter();
       Object payload = e.getPayload();
-      return executeCallback(frame, state, handler, payload, e);
+      return executeCallback(frame, state, panicType, handler, payload, e);
     } catch (AbstractTruffleException e) {
       otherExceptionBranchProfile.enter();
-      Builtins builtins = EnsoContext.get(this).getBuiltins();
-      Object payload = builtins.error().getPolyglotError().wrap(e);
-      return executeCallback(frame, state, handler, payload, e);
+      return executeCallback(frame, state, panicType, handler, e, e);
     }
   }
 
   private Object executeCallback(
       VirtualFrame frame,
       State state,
+      Object panicType,
       Object handler,
       Object payload,
       AbstractTruffleException originalException) {
     Builtins builtins = EnsoContext.get(this).getBuiltins();
     Atom caughtPanic =
         builtins.caughtPanic().getUniqueConstructor().newInstance(payload, originalException);
-    return invokeCallableNode.execute(handler, frame, state, new Object[] {caughtPanic});
+
+    if (profile.profile(isPayloadOfPanicTypeNode.execute(panicType, payload))) {
+      return invokeCallableNode.execute(handler, frame, state, new Object[] {caughtPanic});
+    } else {
+      return throwPanicNode.execute(caughtPanic);
```

Comment:
Hmm, do we need the `throwPanicNode` here?

I'd think in this branch there should be no need to even allocate the `caughtPanic` Atom and we should be able to just `throw interopLibrary.throwException(originalException)`. It should be faster too.

Reply:
Adding a short-circuit that avoids the additional node will work as well. Was trying to avoid DRY.

> It should be faster too.

Whenever exceptions start to float around, one should generally not expect some peak performance.